Add custom element property support behind a flag

josepharhar opened the pull request:
<!--
  Thanks for submitting a pull request!
  We appreciate you spending the time to work on these changes. Please provide enough information so that others can review your pull request. The three fields below are mandatory.

  Before submitting a pull request, please make sure the following is done:

  1. Fork [the repository](https://github.com/facebook/react) and create your branch from `main`.
  2. Run `yarn` in the repository root.
  3. If you've fixed a bug or added code that should be tested, add tests!
  4. Ensure the test suite passes (`yarn test`). Tip: `yarn test --watch TestName` is helpful in development.
  5. Run `yarn test --prod` to test in the production environment. It supports the same options as `yarn test`.
  6. If you need a debugger, run `yarn debug-test --watch TestName`, open `chrome://inspect`, and press "Inspect".
  7. Format your code with [prettier](https://github.com/prettier/prettier) (`yarn prettier`).
  8. Make sure your code lints (`yarn lint`). Tip: `yarn linc` to only check changed files.
  9. Run the [Flow](https://flowtype.org/) type checks (`yarn flow`).
  10. If you haven't already, complete the CLA.

  Learn more about contributing: https://reactjs.org/docs/how-to-contribute.html
-->

## Summary

This implements the solution I've been proposing and discussing for https://github.com/facebook/react/issues/11347 which does an "in" check on the custom element to determine if we should be assigning to the element's property instead of its attribute, as well as adding support for custom events by making JSX attributes which start with "on" event listeners.

This doesn't implement the whenDefined -> reapply properties behavior we have been discussing, I'm planning on addressing that later.
<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->

## Test Plan

I've been keeping [my test site](https://josepharhar.github.io/react-custom-elements/) up to date and this patch includes several automated tests, but I still haven't gotten around to using the `fixtures/attribute-behavior` test to look for more behavior changes yet.

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes the user interface. -->

josepharhar commented on packages/react-dom/src/server/ReactDOMServerFormatConfig.js:
```diff
@@ -1128,12 +1128,17 @@ function pushStartCustomElement(
+      if (enableCustomElementPropertySupport && propKey === 'className') {
+        // className gets rendered as class on the client, so it should be
+        // rendered as class on the server.
+        propKey = 'class';
```
className is special because all elements have a builtin `className` property which triggers the `'className' in node` check when rendering on the client - meaning that JSX on the client will always use a property instead of an attribute for className. Without this, one of the automated tests was getting errors (I think during hydration?) due to this mismatch.
Alternatively we could make it so that when client rendering `className` is always set as an attribute instead of a property, but I think this makes more sense and it's what Preact is already doing.

sebmarkbage replied:
Yea this make sense because htmlFor doesn't have any semantics on custom element which are just Generic HTML elements.

There are a number of other ones that are case sensitive for similar reasons. contentEditable, tabIndex, etc. They probably work if used correctly, but it would be nice to add a warning if they use the wrong case.

There are some that just won't work like `offsetHeight` etc. Since those exist as properties but are read-only. Probably should ensure we have a nice warning about those and why they don't work. So that if they're rendered on the server, it doesn't look like they work.

Similarly there are some we have to be careful with like `innerHTML` and `innerText`. It shouldn't pass through as a property to custom elements - even in production so it can't be a DEV only error. Could be worth a test.

sebmarkbage replied:
I'm surprised innerText and textContent aren't reserved properties in React. Doesn't seem right.

josepharhar replied:
> There are a number of other ones that are case sensitive for similar reasons. contentEditable, tabIndex, etc. They probably work if used correctly, but it would be nice to add a warning if they use the wrong case.

contentEditable is interesting, it looks like the attribute can be either `contentEditable` or `contenteditable`, they both work, and the reflected object property is `contentEditable`.
So I guess if someone writes `<my-custom-element contentEditable={true} />` or `<my-custom-element contenteditable={true} />` they should both work fine, right...?

> There are some that just won't work like offsetHeight etc. Since those exist as properties but are read-only. Probably should ensure we have a nice warning about those and why they don't work. So that if they're rendered on the server, it doesn't look like they work.

I mean, you *can* assign to offsetHeight from script and the browser doesn't complain about it... but yeah I guess it would be weird to have a state where you can assign to an attribute called `offsetHeight` when rendering on the server, but you can't on the client because it will always trigger the `'offsetHeight' in node` check...

Should I consider every global attribute and every built in property?

josepharhar replied:
From offline:
- for camel case reflected attribute/properties like `contentEditable`, we should just have a warning if you server render it as `contenteditable` instead of `contentEditable`. The actual behavior will likely be identical if you choose `contentEditable` or `contenteditable` because in the lowercase form it will assign to the attribute, get reflected to the property, and work - and if you do `contentEditable`, then it will assign to the property, reflect to the attribute, and still work just fine.
- for `innerHTML`, `innerText`, and `textContent`, we should probably just block them like we do for regular elements. If a custom element tries to overload these and needs them in order to function, then maybe that case wouldn't work but I'm not sure it's a very common use case.
  - Sebastian mentioned that there is a list somewhere of blocked properties which has innerHTML and that we should add textContent to it separately from this PR.
- for read only properties like `offsetHeight`, we can just have a dev mode warning. When assigning to these in the browser, it's just a no-op that doesn't throw an error or anything, and it doesn't affect the return value of the property (at least for `offsetHeight`), so we shouldn't be worried about assigning to them.

gaearon replied:
Re: blocklist.

https://github.com/facebook/react/blob/50263d3273b6fc983acc5b0fd52e670399b248b1/packages/react-dom/src/shared/DOMProperty.js#L236-L249

josepharhar replied:
Thanks Dan! I went ahead and added innerText and textContent to that list in this PR behind the flag

josepharhar replied:
I've thought some more about `contentEditable` vs `contenteditable`, and I'm not sure if we should emit any warnings...

I went through all of the global reflected HTML attributes and made a table at the bottom to show their reflected object property name vs the react one, and I found that there are some like `contentEditable` where the React name and the reflected object property are the same, which would result in client rendering passing the `in` check and using the object property, and then there are some like `autofocus`/`autoFocus` where using the React name of `autoFocus` would cause the `in` check to fail and we would end up using the attribute instead...

Basically I don't see the point of a warning based on capitalization because:
1. However you capitalize the JSX attribute its going to end up the same - we don't yet have any evidence to suggest that doing `element.contentEditable = true` is any different from `element.setAttribute('contenteditable', 'true')`
2. There are discrepancies between the React name capitalization and the reflected object property name capitalization - why should we tell developers that they should use the React style capitalization, the object property name capitalization, or the HTML attribute capitalization for custom elements if we aren't doing so already?

HTML attribute (always lowercase) | object property | React JSX attribute name
----|----|----
`class`|`className`|`className`
`contenteditable`|`contentEditable`|`contentEditable`
`accesskey`|`accessKey`|`accessKey`
`autofocus`|`autofocus`|`autoFocus`
`dir`|`dir`|`dir`
`draggable`|`draggable`|`draggable`
`enterkeyhint`|`enterKeyHint`|`enterKeyHint`
`hidden`|`hidden`|`hidden`
`id`|`id`|`id`
`inputmode`|`inputMode`|`inputMode`
`lang`|`lang`|`lang`
`nonce`|`nonce`|`nonce`
`part`|`part`|`part`
`slot`|`slot`|`slot`
`spellcheck`|`spellcheck`|`spellCheck`
`style`|`style`|`style`
`tabindex`|`tabIndex`|`tabIndex`
`title`|`title`|`title`
`translate`|`translate`|`translate`

Writing this table also made me realize something about Dan's initial question about `htmlFor` - `htmlFor` isn't relevant for the special casing for `className` here because custom elements won't get a `htmlFor` object property set by the browser - only built in form control elements have those.

josepharhar replied:
For the read only properties like offsetHeight, I just pushed a commit which emits the warning: https://github.com/facebook/react/pull/22184/commits/1a093e584ce50e2e634aa743e04f9cb8fc2b3f7d

However, the test fails because it throws an exception when assigning to the read only properties - but this exception doesn't occur in the browser or in JSDOM... what kind of browser/DOM setup is being used in these tests?

sebmarkbage replied:
“use strict” throws, right?

josepharhar replied:
Oh wow, today I learned...
What do you think we should do? Remove the test? Make it somehow get rid of the 'use strict'?

sebmarkbage replied:
I believe if you render it on the server and hydrate it wouldn’t error.

The reason for the warning would be so that if you render it on the server, during development you should know not to.

So you can test hydrating it.

josepharhar replied:
Ah, that makes sense!
I replaced the client rendering check with a hydration check and made a test for it.

sebmarkbage commented on packages/react-dom/src/shared/DOMProperty.js:
```diff
@@ -206,6 +206,25 @@ export function getPropertyInfo(name: string): PropertyInfo | null {
+  node: Element) {
+  if (name in (node: any)) {
+    const acceptsBooleans = (typeof (node: any)[name]) === 'boolean';
+    return {
```
This seems a bit slow to do for every property. Likely will need some refactoring. However, is this check also equivalent on the built-ins?

josepharhar replied:
Are you referring to `name in (node: any)` or `(typeof (node: any)[name]) === 'boolean'`?

sebmarkbage commented on packages/react-dom/src/shared/DOMProperty.js:
```diff
@@ -206,6 +206,25 @@ export function getPropertyInfo(name: string): PropertyInfo | null {
+      propertyName: name,
+      attributeName: name,
+      attributeNamespace: null,
+      sanitizeURL: false,
```
I think this is fine if we assume whatever it passes it to is going to sanitize it. However `<a href="javascript:..." is="my-custom-element">` or `<img src="..." is="my-custom-img" />` etc. needs to be sanitized since it might not be upgraded yet. That's an existing issue that we should fix.

josepharhar replied:
Oh my, I didn't even know about the `is` attribute. Does react treat elements with the `is` attribute like custom elements...? If not, then is this fine as is...?

gaearon replied:
>Does react treat elements with the is attribute like custom elements...?

Yes.

josepharhar replied:
Ok, I don't think this should block this PR, but it would be good to follow up on later.
I'll make a note to do so.

sebmarkbage commented on packages/react-dom/src/shared/DOMProperty.js:
```diff
@@ -206,6 +206,25 @@ export function getPropertyInfo(name: string): PropertyInfo | null {
+  name: string,
+  node: Element) {
+  if (name in (node: any)) {
+    const acceptsBooleans = (typeof (node: any)[name]) === 'boolean';
```
Is this what preact does? What's the effective thing here? That if we pass null, we pass a boolean instead of empty string?

Seems like it's wrong either way, because it seems like we should pass null through to the property.

josepharhar replied:
Thats a good point.
I guess the `acceptsBooleans` and `type` flags in PropertyInfo don't really make sense for custom elements, do they...? Should we rethink using PropertyInfo for this feature and instead change all of the logic which uses PropertyInfo?

josepharhar replied:
> Is this what preact does? What's the effective thing here? That if we pass null, we pass a boolean instead of empty string?

I added a "Assignment to boolean properties" section at the bottom of my [test page](https://josepharhar.github.io/react-custom-elements/) for this.

It seems like the only difference is when we try to render the JSX attribute as `null` when the value was previously a boolean: preact will render `null` as an empty string, whereas this patch will render it as `false`.

> Seems like it's wrong either way, because it seems like we should pass null through to the property

I agree that passing `null` here makes more sense, but I'm definitely not completely aware of what use cases are at play here... would you like me to just do `null`? Or ask around why preact does an empty string and ask around which would be better?

gaearon replied:
Is this question still relevant? The code shows up as "Outdated" so I'm not sure.

josepharhar replied:
> Is this question still relevant? The code shows up as "Outdated" so I'm not sure.

Yes - if the custom element's property value is currently `true` or `false`, and then we assign `null` into it through react, then this patch will assign `false` into it. Preact instead does an empty string, which I suppose is even more confusing, but as sebastian pointed out, we should perhaps be passing `null` through instead of `false`...

I'll try to make it just do `null` like sebastian suggested

josepharhar replied:
Ok, I pushed a commit which makes it pass `null` through and also stops using `PropertyInfo` entirely like I mentioned in my earlier comment in this thread.

I think this is resolved now, but it may be a good idea to ask for feedback on this behavior later since I haven't discussed it with anyone outside of this thread.

just-boris commented on packages/react-dom/src/client/DOMPropertyOperations.js:
```diff
@@ -139,15 +141,49 @@ export function setValueForProperty(
+    const alreadyHadListener = (node: any)._listeners && (node: any)._listeners[listenersObjName];
+
+    if (typeof value === 'function' || alreadyHadListener) {
+      if (!(node: any)._listeners) {
```
What if a web-component also uses `_listeners` for its own use-cases?

josepharhar replied:
Great question! This happens to be what preact does in this case, and I *figured* that'd be fine.
Dan and Sebastian, do you have any ideas of where else we could store data for a node in React?

sebmarkbage replied:
It probably should be adding an abstraction here: https://github.com/facebook/react/blob/cae635054e17a6f107a39d328649137b83f25972/packages/react-dom/src/client/ReactDOMComponentTree.js#L39-L44

That uses a random suffix to avoid collisions between multiple Reacts.

I do wonder if it's worth while using a proxy or if we can just pass through. If we can add the raw listener, then we can remove it and checking whether it already exists using `ReactDOMComponentTree.getFiberCurrentPropsFromNode(...)`.

I get that it might be thrashy perf wise if it's always a new listener but a memoized one wouldn't need to.

josepharhar replied:
Thanks for the pointer to those react internals, that looks much better!

> I do wonder if it's worth while using a proxy or if we can just pass through. If we can add the raw listener, then we can remove it and checking whether it already exists using ReactDOMComponentTree.getFiberCurrentPropsFromNode(...).

I see that getFiberCurrentPropsFromNode looks at `__reactProps` on the node. Are you saying that the event listener function is already stored there somehow after we call addEventListener? Or that we should find a place in there to store the event listener?

If I remember correctly, the purpose of the proxy is to be able to quickly re-assign the actual listener function without needing to call addEventListener again because apparently it's faster.

gaearon replied:
For a start, can we just fix this to use a unique prefix with the technique above? And then revisit.

josepharhar commented on packages/react-dom/src/__tests__/DOMPropertyOperations-test.js:
```diff
@@ -155,6 +155,107 @@ describe('DOMPropertyOperations', () => {
+      const container = document.createElement('div');
+      ReactDOM.render(<Test />, container);
+      container.querySelector('span').click();
+      // TODO why doesn't this test pass!? As far as I can tell from outside
```
That did the trick, thanks!
Is this behavior a React thing?

gaearon replied:
I thought that’s how it works in the  browser too.

josepharhar replied:
This code in the browser, which uses a detached element, fires the click event:
```javascript
const div = document.createElement('div');
div.onclick = console.log('click event!');
div.click();
```

josepharhar replied:
Hmm, React typically follows the browser behavior here and allows click events for detached elements: https://jsfiddle.net/jarhar/oujy6bah/
I'm not sure why this test appears to be different.

josepharhar replied:
In any case, the original comment is resolved

josepharhar commented on packages/react-dom/src/__tests__/DOMPropertyOperations-test.js:
```diff
@@ -155,6 +155,107 @@ describe('DOMPropertyOperations', () => {
+    });
+
+    // @gate enableCustomElementPropertySupport
+    it('custom elements should still have onClick treated like regular elements', () => {
```
I removed the gate and I added another assertion to look at event.nativeEvent, which should test that it's a synthetic event.
What do you think?

josepharhar commented on packages/react-dom/src/client/DOMPropertyOperations.js:
```diff
@@ -204,3 +240,11 @@ export function setValueForProperty(
   }
 }
+
+function eventProxy(e: Event) {
```
I changed it to fireEventProxy, how does that look?

josepharhar replied:
I'm guessing `fireEventProxy` is OK

gaearon commented:
Hi @josepharhar! It's been a while since I looked at this, and Seb is going to be out for a bit. Can you clarify what is "done" and what do you need feedback on? I think we don't need to fix every implementation nit at this point, and the important part is to get the tests green and implementation behavior matching what we want. Then we can do another review pass just for implementation. Does this sound like a good plan? Is there anything blocking you from getting to green CI with the implementation? Thanks!

gaearon commented on packages/react-dom/src/__tests__/ReactServerRenderingHydration-test.js:
```diff
@@ -531,4 +533,31 @@ describe('ReactDOMServerHydration', () => {
+        `Warning: Assignment to read-only property will result in a no-op: \`${readOnlyProperty}\``,
+      );
+
+      // Ensure the test fails to satisfy @gate when the flag is turned off and we aren't in development.
```
This seems confusing. I suggest instead to always run the test (if the prod behavior is the same) but to keep the `toErrorDev` in inline `if (gate(...))` checks. So `if (gate(...)) { expect(() => hydrate()).toErrorDev(...) } else { hydrate() }`

josepharhar replied:
That did the trick, thanks!

gaearon commented on packages/react-dom/src/client/DOMPropertyOperations.js:
```diff
@@ -149,9 +150,56 @@ export function setValueForProperty(
+
+    if (typeof value === 'function' || alreadyHadListener) {
+      if (!(node: any)._listeners) {
+        (node: any)._listeners = {};
```
I hid the existing discussion to clean it up. As discussed before, let's change `_listeners` to use a random suffix matching what we do for other fields in DOMComponentTree.

Then we'll look again at the other approach Seb suggested earlier.

josepharhar replied:
Done!
I added a note in my list of followups to consider Seb's approach in the other thread.

gaearon commented on packages/react-dom/src/client/ReactDOMComponent.js:
```diff
@@ -1031,7 +1037,25 @@ export function diffHydratedProperties(
+      ) {
+        // $FlowFixMe - Should be inferred as not undefined.
+        extraAttributeNames.delete(propKey.toLowerCase());
+        if (__DEV__) {
```
This checks seems unnecessary because of the outer one. Or does lint complain?

josepharhar replied:
Yeah, I get this error from `yarn linc` if I remove the `__DEV__` check:
```
/usr/local/google/home/jarhar/react/packages/react-dom/src/client/ReactDOMComponent.js
  1052:9  error  Wrap console.error() in an "if (__DEV__) {}" check  react-internal/no-production-logging

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.
```

josepharhar commented:
> Hi @josepharhar! It's been a while since I looked at this, and Seb is going to be out for a bit. Can you clarify what is "done" and what do you need feedback on? I think we don't need to fix every implementation nit at this point, and the important part is to get the tests green and implementation behavior matching what we want. Then we can do another review pass just for implementation. Does this sound like a good plan? Is there anything blocking you from getting to green CI with the implementation? Thanks!

Thanks so much! I agree, and I don't have anything blocking me at this point. Here is a list of followups I've created from the comment threads:
- ask for feedback from users/developers on boolean properties: https://github.com/facebook/react/pull/22184#discussion_r697777188
- address URL sanitization comment: https://github.com/facebook/react/pull/22184#discussion_r697775743
- consider if we should be using an event proxy function like preact does or not https://github.com/facebook/react/pull/22184#discussion_r698816725

josepharhar commented:
@gaearon Do you have any more thoughts? Is this ready to merge?

gaearon commented:
I'll block some time next week to review.

gaearon commented:
Are there projects where we want to test it? When we merge this will go out in the `@experimental` releases.

josepharhar commented:
> When we merge this will go out in the @experimental releases.

Nice! I was wondering how I would instruct people to test it out.

> Are there projects where we want to test it?

I don't have any projects in mind. I was just planning on posting back in the github issue to ask people to test it out.

josepharhar commented:
> Just throwing this idea out there: you could potentially test it against the Custom Elements Everywhere test script and see what the output is? https://github.com/webcomponents/custom-elements-everywhere

Here is the output with the current state of this PR (Disclaimer that I couldn't get the main branch of custom-elements-everywhere to build, this is based on a commit in that repo 11 months ago):
<img width="753" alt="Screen Shot 2021-11-16 at 12 07 27 PM" src="https://user-images.githubusercontent.com/7681467/142041306-b00dd8d6-e366-4a6a-95ab-025910a15d34.png">

developit commented on packages/react-dom/src/client/ReactDOMComponent.js:
```diff
@@ -1031,7 +1037,25 @@ export function diffHydratedProperties(
+      } else if (
+        enableCustomElementPropertySupport &&
+        isCustomComponentTag &&
+        (propKey === 'offsetParent' ||
+          propKey === 'offsetTop' ||
+          propKey === 'offsetLeft' ||
+          propKey === 'offsetWidth' ||
+          propKey === 'offsetHeight' ||
+          propKey === 'isContentEditable')
+      ) {
```
I think there are quite a few more properties that throw when assigned to, aren't there?
A quick test indicates you may want to add at least `outerText` and `outerHTML`.

josepharhar replied:
Thanks! I just pushed a commit to include `outerText` and `outerHTML`. Fortunately this is just a warning, so if it's not perfect we can always change it later.

matthewp commented on packages/react-dom/src/client/DOMPropertyOperations.js:
```diff
@@ -149,9 +151,52 @@ export function setValueForProperty(
+
+  if (
+    enableCustomElementPropertySupport &&
+    isCustomComponentTag &&
```
Does this need to be restricted to just custom components? It's possible to trigger custom events on any element and there are many reasons to do so.

josepharhar replied:
I know that's what preact does, but I'm worried about breaking existing usage with regular elements. There was already significant concern with how this reserves everything starting with "on" for custom events, like anything starting with "one" and "once" for example.

matthewp commented on packages/react-dom/src/client/DOMPropertyOperations.js:
```diff
@@ -149,9 +151,52 @@ export function setValueForProperty(
+
+  if (
+    enableCustomElementPropertySupport &&
+    isCustomComponentTag &&
```
This presumably means this won't work with extended builtins.

If there isn't a good reason for doing so, I would recommend treating custom elements and non-custom elements the same and only special case where they are truly different/special. This will lead to fewer bugs and fewer places where the code diverges into multiple paths.

josepharhar replied:
I'm not sure what you mean by extended builtins, but using properties instead of attributes for regular elements in addition to custom elements was never considered before while I've been working on this and sounds like a much bigger change than what I've gotten agreement on so far.

matthewp replied:
You can extend builtin elements like `<button>` which will not have a dasherized name and not be `isCustomComponentTag`. https://developers.google.com/web/fundamentals/web-components/customelements#extendhtml

It's not supported in Safari so not a big deal to not support here, I was just using it as an example of how all of the `isCustomComponentTag` special-casing is going to cause some use-cases to not be covered.

Agree that there is a larger issue to discuss, in another forum. Less special-casing would lead to more uniformity of implementation (easier to maintain) and be easier for users to understand since everything would work the same way.

josepharhar replied:
Ah I see, thanks for the clarification! `isCustomComponentTag` appears to include elements with the `is` attribute from the link you shared: https://github.com/facebook/react/blob/b831aec48f2d54913b836d2288b62b7eee9fbca1/packages/react-dom/src/shared/isCustomComponent.js#L12

I am happy to continue this discussion elsewhere as well.

MarcusNotheis commented:
> Are there projects where we want to test it? When we merge this will go out in the `@experimental` releases.

[SAP/ui5-webcomponents-react](https://github.com/SAP/ui5-webcomponents-react) would definitely be in for a test as our project is heavily based on custom elements :)

gaearon commented:
Sorry, spent the whole week working on ReactConf preparations. Next week it is.

gaearon commented:
Is there an up-to-date document that describes the final strategy and a spec of what we do in each case?

gaearon commented on packages/react-dom/src/client/DOMPropertyOperations.js:
```diff
@@ -149,9 +151,52 @@ export function setValueForProperty(
+    let eventName = name.replace(/Capture$/, '');
+    const useCapture = name !== eventName;
+    const nameLower = eventName.toLowerCase();
+    if (nameLower in node) {
```
Removing this `if` block doesn't fail any tests. What is it for? Do we need a test for coverage?

josepharhar replied:
I copied this over from these lines in preact: https://github.com/preactjs/preact/blob/dd1e281ddc6bf056aa6eaf5755b71112ef5011c5/src/diff/props.js#L88-L90

I suppose it's to make it so `onClick` listens for an event named `click` instead of `Click`, but only if the element has an `onclick` property. The comment suggests that this only matters for builtin elements, and since removing this line still makes the `<my-custom-element onClick={...} />` test pass, it can be removed. I'll remove it.

gaearon commented on packages/react-dom/src/client/DOMPropertyOperations.js:
```diff
@@ -149,9 +151,52 @@ export function setValueForProperty(
+    const listeners = getCustomElementEventHandlersFromNode(node);
+    const alreadyHadListener = listeners[listenersObjName];
+
+    if (typeof value === 'function' || alreadyHadListener) {
```
Hmm. So we do different things depending on the value type. This makes me a bit uneasy. I guess that's existing practice in Preact etc?

This makes me uneasy because props can always change midway. E.g. you get function on first render and something else on the second render. Do we have tests demonstrating what exactly happens when the type changes? The guarantee we want to preserve is that A -> B -> C -> D should have the same "end result" as just D, regardless of what A, B, C were. E.g. number -> string -> function -> number, or number -> function -> function -> string.

gaearon replied:
Here's a concrete example.

This doesn't throw:

```js
      const container = document.createElement('div');
      // ReactDOM.render(<Test handler={oncustomevent} />, container);
      ReactDOM.render(<Test handler="hello" />, container);
      const customElement = container.querySelector('my-custom-element');
      customElement.dispatchEvent(new Event('customevent'));
```

but this throws:

```js
      const container = document.createElement('div');
      ReactDOM.render(<Test handler={oncustomevent} />, container);
      ReactDOM.render(<Test handler="hello" />, container);
      const customElement = container.querySelector('my-custom-element');
      customElement.dispatchEvent(new Event('customevent'));
```

>TypeError: (0 , _ReactDOMComponentTree.getCustomElementEventHandlersFromNode)(...)[(e.type + "false")] is not a function

This doesn't seem like consistent behavior.

josepharhar replied:
Thanks for pointing out this case! I like the fully consistent behavior option.

More on the function type behavior - Preact doesn't look at the type of the value passed in and unconditionally forwards it to addEventListener. Jason Miller endorsed this behavior by telling me that it supports the [EventListener interface](https://developer.mozilla.org/en-US/docs/Web/API/EventListener) better because addEventListener can take objects in addition to just functions, and that nobody has ever complained about all properties with `on` being reserved.

I know that Sebastian seemed objected to reserving everything that starts with `on` (meaning you can't do `one={'foo'}` for example). I believe that this behavior of looking for `function`s strongly mitigates this issue, and I don't really see the EventListener interface being a big issue either. However, I don't exactly remember what Sebastian's last thoughts about this were, and I don't remember exactly when I decided to start looking for the `function` type. I did propose the behavior [in this comment](https://github.com/facebook/react/issues/11347#issuecomment-887889479), and there wasn't really anyone disagreeing.

In the end, I think that we should go forward with this and do the Fully consistent behavior option. I'll plan on implementing it once I get through the rest of the comments in this PR.

josepharhar replied:
I added some code here to implement the fully consistent option - when calling addEventListener, it will now try to figure out if we previously assigned the same prop as an attribute or property, in which case it will assign null into it.

gaearon commented on packages/react-dom/src/client/DOMPropertyOperations.js:
```diff
@@ -149,9 +151,52 @@ export function setValueForProperty(
+    name[1] === 'n'
+  ) {
+    let eventName = name.replace(/Capture$/, '');
+    const useCapture = name !== eventName;
```
I can hardcode this to `true` and tests still pass. We should have test coverage for `Capture` events firing in the capture phase. We also need test coverage for both bubble + capture listeners on the same element, to verify that they don't clash due to some mistake.

josepharhar replied:
Thanks for spotting this! I wrote a test but it was broken :(
I fixed it and now it will fail when useCapture is hard coded to true or false.

gaearon commented on packages/react-dom/src/__tests__/DOMPropertyOperations-test.js:
```diff
@@ -155,6 +155,186 @@ describe('DOMPropertyOperations', () => {
+      expect(customElement).toBe(customElement);
+      customElement.dispatchEvent(new Event('customevent'));
+      expect(oncustomevent).toHaveBeenCalledTimes(1);
+    });
```
We should also verify that we can re-attach a function again. Additionally, we should have coverage that we're able to attach different functions over time, and the correct one fires (while others don't).

josepharhar replied:
Done, thanks!

gaearon commented on packages/react-dom/src/client/DOMPropertyOperations.js:
```diff
@@ -149,9 +151,52 @@ export function setValueForProperty(
+    isCustomComponentTag &&
+    name in (node: any)
+  ) {
+    (node: any)[name] = value;
```
We should have test coverage verifying that this _does not_ get set in cases where we don't expect. E.g. if `oncustomevent` is _not_ supposed to set the `oncustomevent` property even when it exists, we should have a test verifying this. That test should specify what happens when the value type changes midway (e.g. function -> string and back).

josepharhar replied:
Thanks for pointing this out!
I added some tests with a variety of switching between nothing, string, and function.

gaearon commented on packages/react-dom/src/client/DOMPropertyOperations.js:
```diff
@@ -216,3 +261,11 @@ export function setValueForProperty(
   }
 }
+
+function fireEventProxy(e: Event) {
```
Coming back to what I think @sebmarkbage what suggesting — how do you feel about this alternative approach? Am I missing something?

```diff
@ packages/react-dom/src/client/DOMPropertyOperations.js:25 @ import {
  enableCustomElementPropertySupport,
} from 'shared/ReactFeatureFlags';
import {checkAttributeStringCoercion} from 'shared/CheckStringCoercion';
import {getCustomElementEventHandlersFromNode} from './ReactDOMComponentTree';
import {getFiberCurrentPropsFromNode} from './ReactDOMComponentTree';

import type {PropertyInfo} from '../shared/DOMProperty';

@ packages/react-dom/src/client/DOMPropertyOperations.js:169 @ export function setValueForProperty(
    }
    eventName = eventName.slice(2);

    const listenersObjName = eventName + (useCapture ? 'true' : 'false');
    const listeners = getCustomElementEventHandlersFromNode(node);
    const alreadyHadListener = listeners[listenersObjName];
 
    if (typeof value === 'function' || alreadyHadListener) {
      listeners[listenersObjName] = value;
      const proxy = useCapture ? fireEventProxyCapture : fireEventProxy;
      if (value) {
        if (!alreadyHadListener) {
          node.addEventListener(eventName, proxy, useCapture);
        }
      } else {
        node.removeEventListener(eventName, proxy, useCapture);
      }
    const prevProps = getFiberCurrentPropsFromNode(node);
    const prevValue = prevProps != null ? prevProps[name] : null;
    if (typeof prevValue === 'function') {
      node.removeEventListener(eventName, prevValue, useCapture)
    }
    if (typeof value === 'function') {
      node.addEventListener(eventName, value, useCapture);
      return;
    }
  }
@ packages/react-dom/src/client/DOMPropertyOperations.js:258 @ export function setValueForProperty(
  }
}

function fireEventProxy(e: Event) {
  getCustomElementEventHandlersFromNode(this)[e.type + 'false'](e);
}
 
function fireEventProxyCapture(e: Event) {
  getCustomElementEventHandlersFromNode(this)[e.type + 'true'](e);
}
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
modified: packages/react-dom/src/client/ReactDOMHostConfig.js
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
@ packages/react-dom/src/client/ReactDOMHostConfig.js:445 @ export function commitUpdate(
  newProps: Props,
  internalInstanceHandle: Object,
): void {
  // Apply the diff to the DOM node.
  updateProperties(domElement, updatePayload, type, oldProps, newProps);
  // Update the props handle so that we know which props are the ones with
  // with current event handlers.
  updateFiberProps(domElement, newProps);
  // Apply the diff to the DOM node.
  updateProperties(domElement, updatePayload, type, oldProps, newProps);
}
```

This lets us get rid of the proxies. Note that this shouldn't run at all for memoized handlers because they should get filtered out at `prepareUpdate` level and we shouldn't get into `setsetValueForProperty` for them at all.

josepharhar replied:
This is great! I have committed it. I didn't know what getFiberCurrentPropsFromNode does, this is much better!

gaearon commented on packages/react-dom/src/__tests__/DOMPropertyOperations-test.js:
```diff
@@ -155,6 +155,186 @@ describe('DOMPropertyOperations', () => {
+      expect(customElement.getAttribute('onobj')).toBe('[object Object]');
+      expect(customElement.getAttribute('onarray')).toBe('one,two');
+      expect(customElement.getAttribute('ontrue')).toBe('true');
+      expect(customElement.getAttribute('onfalse')).toBe('false');
```
Let's dispatch the corresponding event names here to verify nothing crashes.

josepharhar replied:
done

gaearon commented on packages/react-dom/src/__tests__/DOMPropertyOperations-test.js:
```diff
@@ -155,6 +155,186 @@ describe('DOMPropertyOperations', () => {
+      ReactDOM.render(<my-custom-element innerText="foo" />, container);
+      const customElement = container.querySelector('my-custom-element');
+      expect(customElement.getAttribute('innerText')).toBe(null);
+      expect(customElement.hasChildNodes()).toBe(false);
```
For all tests like this, let's add one more update with a different value (and a series of checks) to verify the update codepath doesn't accidentally let something through.

josepharhar replied:
done

gaearon commented on packages/react-dom/src/server/DOMMarkupOperations.js:
```diff
@@ -30,12 +31,33 @@ import quoteAttributeValueForBrowser from './quoteAttributeValueForBrowser';
-export function createMarkupForProperty(name: string, value: mixed): string {
-  const propertyInfo = getPropertyInfo(name);
-  if (name !== 'style' && shouldIgnoreAttribute(name, propertyInfo, false)) {
+export function createMarkupForProperty(
```
This function is inlined into `pushAttribute` in `ReactDOMServerFormatConfig`. So I'd expect them to have to change in sync. Is there some reason why that wasn't needed?

And if it _was_ needed, then do we know why none of the tests have failed despite missing logic there? This seems to point to missing coverage.

gaearon replied:
So there's two things here.

- It appears this codepath doesn't run anymore at all. This is likely because `ReactPartialRenderer` isn't used in open source at all anymore. It's only used at FB while we haven't switched on new SSR engine yet. So we should probably be careful here before merge. I'll check if we can disable the old renderer so that we're not scared to merge without test coverage.

- The codepath in `ReactDOMServerFormatConfig` _does_ run — but it doesn't have these changes (so it's probably wrong). And because we didn't have a test written specifically to verify these changes, we missed this. So we need to add a test that verifies all of the SSR behavior we care about. You can either use our SSR integration suite (and add more cases there, potentially copy-pasting the Attributes file into a separate one focused on Custom Elements) or you can add more "manual" cases to `ReactServerRendering-test.js` or nearby. But we should make sure to have a regression test for all the design decisions we've made about the output.

josepharhar replied:
Thanks for catching this!
I removed the changes to DOMMarkupOperations and implemented the proposed SSR changes in ReactDOMServerFormatConfig and added tests in ReactServerRendering-test.

gaearon reviewed (changes requested):
Thanks for the work so far!

- Let's add more test coverage as mentioned in individual review comments:
  - Cases where listeners are added, updated, removed. Ensure only latest fires.
  - Cases where value type changes midway. Document the current behavior via tests.
  - Verify we're not setting properties for cases where we shouldn't.
  - Missing SSR test coverage.
- Changes to functionality:
  - Port changes to new SSR.
  - Fix crash when changing value type.
  - Consider the `getCurrrentFiberProps` approach to simplify the "proxying" thing.

In general, they to ensure that:

- Changing any line fails *some* test. This ensures we know what's intentional or not, and avoids regressions when refactoring or changing approaches.
- Any intentional design decisions ("X behaves in a Y way") has a test associated with it. So that we can distinguish intentional decisions and bugs.

We're okay with having many tests.

Thank you!

zlumer commented:
I just wanted to say a huge thank you to @josepharhar for the amount of work put into this PR.

josepharhar commented:
> Is there an up-to-date document that describes the final strategy and a spec of what we do in each case?

This is the last document I made: https://docs.google.com/document/d/1wwL_YZ5TnYorEEr2Lkeh9rgbnsynDXXI7tCejVcpE6I/edit#
There is one behavior that has diverged from this proposal: the usage of the type of the variable when deciding to use addEventListener or not. I originally proposed that we do the same thing as preact here (forward everything starting with "on" to addEventListener), but if I remember correctly Sebastian felt strongly that we shouldn't reserve everything starting with "on" for event listeners. Jason Miller also felt strongly about this the other way, so maybe it's something that I can revisit/discuss more after merging this PR - maybe more users will chime in if they get to test it out.

I added another test for the behavior I discussed in the doc about not removing attributes when assigning to properties.

> Missing SSR test coverage. We need to understand the exact current behavior, including edge cases. (E.g. that we don't currently set properties on hydration.)

I added some tests for this, thanks!

> Changing any line fails some test. This ensures we know what's intentional or not, and avoids regressions when refactoring or changing approaches.
> Any intentional design decisions ("X behaves in a Y way") has a test associated with it. So that we can distinguish intentional decisions and bugs.

I skimmed the patch as well as the doc, and I believe everything is being tested now

trusktr commented:
This is not flexible enough for custom element users (https://github.com/facebook/react/issues/11347#issuecomment-1732205832).

In today's modern Custom Elements world, users need choice:

- Sometimes users will want the default behavior of this PR (JS properties if detected, fallback to attributes)
- Sometimes, users always want to use an attribute regardless if the custom element supports both, for styling purposes (`some-element[foo="bar"] { ... }`).
- Quite often a custom element author writes elements that map attributes to JS properties, but despite that, and despite the JS property existing, users need to set the attribute in order to trigger the custom element's own internal style (ShadowRoot contains `:host[some-prop] { ... }`). This is a requirement set by custom element authors.
- Sometimes an element does not have a property, but the user needs to set properties because something outside of the custom element is reading them as data.
- Sometimes custom element authors must set attributes regardless because something outside of the custom elements is reading them as data (HTMX, Alpine.js, etc)

This is why Lit's [`html`](https://lit.dev/docs/templates/expressions/) (cc @justinfagnani), Solid's [JSX](https://www.solidjs.com/guides/rendering#attributes-and-props), Solid's [`html`](https://github.com/solidjs/solid/tree/main/packages/solid/html) (cc @ryansolid), and others, were designed to allow these use cases be explicitly controllable by the user.

I've experienced all of these use cases _in a single app_. They are not rare.

Here's a Lit example:

```js
return html`
  <some-element
    ?some-boolean=${this.someBoolean}
    another-booolean
    some-attribute=${this.someString}
    .some-property=${this.someNonString}
  ></some-element>
`
```

Solid uses `attr:foo=` and `prop:foo=` syntax in both JSX and `html` to give users control.


For the above reasons, this solution does not meet the requirements that custom element users have, and this solution is incomplete by today's standards.

Custom element users need a _complete solution_ to this. If not with syntax, then with something runtime like "directives". F.e. `foo={attr(someValue)}` setting an attribute vs `foo={prop(someValue)}` setting a JS prop.

trusktr commented:
@jfbrennan since [you were curious](https://github.com/facebook/react/issues/11347#issuecomment-1742370624), see that ☝️. Custom element users must be responsible for knowing how custom elements are used, including whether to set an attribute or a property, and no system can accurately determine this. We're still gonna need to resort to `useRef` and imperative code, missing out on declarative JSX benefits for cases not covered.

jfbrennan commented:
That's too bad, but actually a good thing long term. Proprietary frameworks that consider themselves the center of the universe will die, those that have already caught up with the needs of the WC era will live on.

jmleep commented:
Maybe I'm making a wrong assumption, but I installed `react@experimental` and `react-dom@experimental` (looking at [this guidance](https://react.dev/reference/react-dom/components#custom-html-elements) on docs), but I couldn't get any of my custom events to fire by prefixing the event name with `on`. It looks like this is scheduled to go out in the next major release, is that still correct?

josepharhar commented:
Can you provide a minimal example?

Amrender-Singh commented:
Hi React Team,

I'm reporting a potential bug I encountered while testing custom elements with the latest versions of react@experimental and react-dom@experimental.

Observed behavior:

I created a custom element with a callback function bound to a custom event.
When the event is triggered on the custom element, the callback does not fire.

Example:
https://stackblitz.com/edit/vitejs-vite-vpqkrw?file=src%2FApp.jsx

When I attach the event listener directly from DOM, the callback is getting triggered.

eps1lon commented:
@Amrender-Singh I can't get this to work with plain HTML either: https://stackblitz.com/edit/vitejs-vite-vuvkyc?file=index.html,customElement.js


I believe the problem is that you're dispatching the custom event at `this` which is not the instance of the custom element. There's a question on StackOverflow (https://stackoverflow.com/a/70789592) that describes the same problem I believe.

eps1lon commented:
@trusktr The cases you believe are currently not covered are best raised in a new issue for visibility. Though I believe they can be covered by how React implements custom elements in a way that seems to me more like idiomatic HTML:

>Sometimes, users always want to use an attribute regardless if the custom element supports both, for styling purposes (some-element[foo="bar"] { ... }).

Isn't the solution here either using [Custom states and custom state pseudo-class CSS selectors](https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_custom_elements#custom_states_and_custom_state_pseudo-class_css_selectors) or attribute reflection? Just like you must use `:checked` instead of `[checked]` to style an HTML checkbox, you'd have to use `some-element:state(foo)`. For the cases where you need `[foo="bar"]` wouldn't you have implemented attribute reflection for the property in your custom element anyway? 

> Sometimes an element does not have a property, but the user needs to set properties because something outside of the custom element is reading them as data.

You'd have to use a ref here for built-in browser components as well, no? Seems like this is a feature request that's not specific to custom elements.

Amrender-Singh commented:
@eps1lon It is working fine in HTML+JS, the main  issue with your example was you were not listening to the event correctly, I  think in order  to listen for custom events using vanilla JS you need to register listener using  addEventListener.

https://open-wc.org/guides/knowledge/events/
https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener


Example: 
https://plnkr.co/edit/UT2jGd67TA6ePyiH?open=lib%2Fscript.js

eps1lon commented:
Not sure if we should special case the first letter here and map `onSomecustomevent` to `somecustomevent`. That's now how you'd write in HTML but then you'd also not write `onClick` either. Maybe it's better to break the capitalisation magic early so that people don't assume that we can magically also map `onCustomClick` to `customclick`.

The main confusion here was that React removes the `on` from the event listener not that the listener is case-sensitive.

jmleep commented:
How does this work with events that use dashes or colons? 

`custom-event`

eps1lon commented:
> How does this work with events that use dashes or colons?
> 
> `custom-event`

events with dashes already have a test: https://github.com/facebook/react/pull/22184/files#diff-db98d094dc4f5b1a9c32b4da4df310370707408a929762868d44bf1acd515923R188

events with colon do not work as JSX attribute due to not being supported by React's JSX though we don't have a test if they're spread to props as an object e.g. `<div {...{ 'oncustom:click': () => console.log(1) }} />`. @jmleep could you add a test to `packages/react-dom/src/__tests__/DOMPropertyOperations-test.js` just like for the other event handler test but with `<div {...{ 'oncustom:click': () => console.log(1) }} />` instead?

trusktr commented:
> Isn't the solution here either using [Custom states and custom state pseudo-class CSS selectors](https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_custom_elements#custom_states_and_custom_state_pseudo-class_css_selectors) or attribute reflection?

That's one possible solution that a custom element *author* (not the user) has.

> For the cases where you need `[foo="bar"]` wouldn't you have implemented attribute reflection for the property in your custom element anyway?

As a custom elements *user* (not an author) there is no guarantee that every author of all custom elements that ever exist will follow any rules specifically for React.

Because of this, any templating system that exists (not only React JSX) and supports custom elements needs to give users the ability to choose whether they set HTML attributes vs JS properties.

[Lit's `html`](https://lit.dev/docs/templates/overview/) template tag is the reference implementation for this type of support, as it was designed for custom elements from the start. [Solid's `html`](https://github.com/solidjs/solid/tree/main/packages/solid/html) template tag follows in second place, currently missing boolean attribute syntax.

See here for a better description of the problem and how Lit and Solid support use cases:

- https://github.com/solidjs/solid/issues/2158

And see here for the `custom-elements-everywhere` test that will document support or lack of support for this across frameworks:

- https://github.com/webcomponents/custom-elements-everywhere/issues/2352

---

> > Sometimes an element does not have a property, but the user needs to set properties because something outside of the custom element is reading them as data.
> 
> You'd have to use a ref here for built-in browser components as well, no? Seems like this is a feature request that's not specific to custom elements.

Sometimes you just need to set a JS property and not an attribute, for *whatever reason* (you might not need a ref at all, you just want to set the JS property in the JSX, and something else like a jQuery plugin outside of the component could be reading the JS properties).

"Use a ref" essentially means "just use plain JS (instead of React JSX)". Plain JS has always worked, but that defeats the goal here, which is to not have to break out of React JSX into plain JS (because that's much more convenient for the developer experience).
